// Per business logic asset = lamppost

syntax = "proto3";

package lighting.asset;

import "location.proto";

service Asset {
    rpc Get (Request) returns (Reply);

    rpc List (ListRequest) returns (stream ListReply);

    rpc Create (Reply) returns (Reply);

    rpc Update (Reply) returns (Reply);

    rpc Delete (Request) returns (Reply);

    rpc Prune (Request) returns (Reply);
}

message ListRequest {
    int32 limit = 1;
    int32 offset = 2;
}

message Request {
    int32 id = 1;
}

message Reply {
    int32 id = 1;
    ActivityStatus status = 2;

    // send UIDs of Elements connected to this asset.
    // can't send element itself because that would introduce circular imports and gRPC throws up.
    repeated int32 element_uids = 3;

    // where is this asset?
    oneof location_oneof {

        // if setting a location, leave this unset - default value is False
        bool no_location = 14;

        // if no_location is False, then read this location. Else, leave this alone.
        // Deals with edge case where there are elements at exactly (0, 0), (x, 0) and (0, y).
        lighting.location.Location location = 15;
    }
}

message ListReply {
    repeated Reply assets = 1;
}

enum ActivityStatus {
    UNAVAILABLE = 0;
    ACTIVE = 1;
    INACTIVE = 2;
    DELETED = 15;
}
